File: SchoolERP/bin/Release/Publish/Angular/Admin/Controller/StudentRegistrationController.js
(function () {
    "use strict";

    angular.module("app").controller("StudentRegistrationController", StudentRegistrationController);

    StudentRegistrationController.$inject = ["$scope", "adminService", "loadDataService", "ConstantData"];

    function StudentRegistrationController($scope, adminService, loadDataService, ConstantData) {
        $scope.checkSysytemLogin();
        $scope.GenderList = ConstantData.Gender;
        $scope.PaymentModeList = ConstantData.PaymentMode;
        $scope.StateList = [];
        $scope.PermanentCityList = [];

        $scope.resetRegistration = function () {
            $scope.Registration = {};
            $scope.Registration.StudentTypeId = null;
            $scope.Registration.Gender = null;
            $scope.Registration.ClassId = null;
            $scope.Registration.AcademicSessionId = null;
            $scope.Registration.PermanentCityId = null;
            $scope.Registration.PermanentStateId = null;
            $scope.Registration.PaymentMode = null;
            $scope.Registration.PaymentDate = new Date();
            $scope.Registration.DemandDraftDate = new Date();
            $scope.Registration.ChequeDate = new Date();
            $scope.Registration.RegistrationDate = new Date();
            $scope.RegistrationFeeList = [];
            $scope.IsSubmitted = false;
        }
        $scope.resetRegistration();

        $scope.$watch('formRegistration.$valid', function (value) {
            $scope.IsFormValid = value;
        })

        $scope.getClassList = function () {
            $scope.dataLoading = true;
            adminService.getClassList(ConstantData.ClassStatus[1].Key)
                .then(function (response) {
                    if (response.data.Message == 'Success') {
                        $scope.ClassList = response.data.ClassList;
                        var index = { ClassId: null, ClassName: '--Select Class--' };
                        $scope.ClassList.unshift(index);
                    } else {
                        toastr.error(response.data.Message);
                    }
                    $scope.dataLoading = false;
                }, function (err) {
                    toastr.error("Some error occured while fetching data");
                    $scope.dataLoading = false;
                })
        }
        $scope.getClassList();

        $scope.changePermanentState = function () {
            $scope.PermanentCityList = [];
            angular.forEach($scope.CityList, function (city, key) {
                if (city.StateId == $scope.Registration.PermanentStateId) {
                    $scope.PermanentCityList.push(city);
                }
            })
            var index = { CityId: null, CityName: '--Select City--' };
            $scope.PermanentCityList.unshift(index);
        }

        $scope.getCityList = function () {
            $scope.dataLoading = true;
            adminService.getCityList(ConstantData.CityStatus[1].Key)
                .then(function (response) {
                    if (response.data.Message == 'Success') {
                        $scope.CityList = response.data.CityList;
                        $scope.changePermanentState();
                    } else {
                        toastr.error(response.data.Message);
                    }
                    $scope.dataLoading = false;
                }, function (err) {
                    toastr.error("Some error occured while fetching data");
                    $scope.dataLoading = false;
                })
        }
        $scope.getCityList();

        $scope.getStateList = function () {
            $scope.dataLoading = true;
            adminService.getStateList(ConstantData.StateStatus[1].Key)
                .then(function (response) {
                    if (response.data.Message == 'Success') {
                        $scope.StateList = response.data.StateList;
                        var index = { StateId: null, StateName: '--Select State--' };
                        $scope.StateList.unshift(index);
                    } else {
                        toastr.error(response.data.Message);
                    }
                    $scope.dataLoading = false;
                }, function (err) {
                    toastr.error("Some error occured while fetching data");
                    $scope.dataLoading = false;
                })
        }
        $scope.getStateList();

        $scope.getAcademicSessionList = function () {
            if ($scope.Registration.ClassId == null || $scope.Registration.ClassId == undefined) {
                $scope.AcademicSessionList = [];
                $scope.AcademicSessionId = null;
                return;
            }
            var obj = {
                ClassId: $scope.Registration.ClassId,
                Status: 1
            }
            $scope.dataLoading = true;
            adminService.getAcademicSessionList(obj)
                .then(function (response) {
                    if (response.data.Message == 'Success') {
                        $scope.AcademicSessionList = response.data.AcademicSessionList;
                    } else {
                        toastr.error(response.data.Message);
                    }
                    $scope.dataLoading = false;
                }, function (err) {
                    toastr.error("Some error occured while fetching data");
                    $scope.dataLoading = false;
                })
        }
        $scope.getAcademicSessionList();

        $scope.getStudentTypeList = function () {
            $scope.dataLoading = true;
            adminService.getStudentTypeList(ConstantData.StudentTypeStatus[1].Key)
                .then(function (response) {
                    if (response.data.Message == 'Success') {
                        $scope.StudentTypeList = response.data.StudentTypeList;
                        var index = { StudentTypeId: null, TypeName: '--Select Student Type--' };
                        $scope.StudentTypeList.unshift(index);
                    } else {
                        toastr.error(response.data.Message);
                    }
                    $scope.dataLoading = false;
                }, function (err) {
                    toastr.error("Some error occured while fetching data");
                    $scope.dataLoading = false;
                })
        }
        $scope.getStudentTypeList();

        $scope.calculateTotalRegistrationFee = function () {
            $scope.TotalRegistrationFee = 0;
            angular.forEach($scope.RegistrationFeeList, function (registrationFee, Key) {
                $scope.TotalRegistrationFee += registrationFee.Amount;
            })
        }

        $scope.getRegistrationFeeList = function () {
            if ($scope.Registration.AcademicSessionId > 0 && $scope.Registration.ClassId > 0 && $scope.Registration.StudentTypeId > 0) {
                $scope.dataLoading = true;
                adminService.getRegistrationFeeListForRegistration($scope.Registration.ClassId, $scope.Registration.StudentTypeId, $scope.Registration.AcademicSessionId)
                    .then(function (response) {
                        if (response.data.Message == 'Success') {
                            $scope.RegistrationFeeList = response.data.RegistrationFeeList;
                            $scope.calculateTotalRegistrationFee();
                            if ($scope.RegistrationFeeList.length == 0) {
                                toastr.error("No registration fee head was found.");
                            }
                        } else {
                            toastr.error(response.data.Message);
                        }
                        $scope.dataLoading = false;
                    }, function (err) {
                        toastr.error("Some error occured while fetching data");
                        $scope.dataLoading = false;
                    })
            } else {
                $scope.RegistrationFeeList = [];
                $scope.TotalRegistrationFee = 0;
            }
        }


        $scope.saveStudentRegistration = function () {
            $scope.IsSubmitted = true;
            if ($scope.IsFormValid) {
                if ($scope.RegistrationFeeList.length == 0) {
                    toastr.error("No registration fee head was found!!");
                    return;
                }

                var saveStudentRegistrationModel = {
                    StudentRegistration: {
                        StudentTypeId: $scope.Registration.StudentTypeId,
                        AcademicSessionId: $scope.Registration.AcademicSessionId,
                        ClassId: $scope.Registration.ClassId,
                        Remarks: $scope.Registration.Remarks,
                        AadharNo: $scope.Registration.AadharNo,
                        Address: $scope.Registration.Address,
                        PermanentCityId: $scope.Registration.PermanentCityId,
                        DOB: loadDataService.getDateTime($scope.Registration.DateOfBirth),
                        EmailId: $scope.Registration.EmailId,
                        FatherName: $scope.Registration.FatherName,
                        Name: $scope.Registration.FullName,
                        Gender: $scope.Registration.Gender,
                        MobileNo: $scope.Registration.MobileNo,
                        MotherName: $scope.Registration.MotherName,
                        RegistrationDate: loadDataService.getDateTime($scope.Registration.RegistrationDate),
                    },
                    FeeCollection: {
                        PaymentDate: loadDataService.getDateTime($scope.Registration.PaymentDate),
                        PaymentMode: $scope.Registration.PaymentMode,
                        Remarks: $scope.Registration.PaymentRemarks,
                        ChequeNo: $scope.Registration.ChequeNo,
                        ChequeDate: loadDataService.getDateTime($scope.Registration.ChequeDate),
                        ChequeBank: $scope.Registration.ChequeBank,
                        ChequeBankBranch: $scope.Registration.ChequeBankBranch,
                        DemandDraftNo: $scope.Registration.DemandDraftNo,
                        DemandDraftDate: loadDataService.getDateTime($scope.Registration.DemandDraftDate),
                        DemandDraftBank: $scope.Registration.DemandDraftBank,
                        TransactionNo: $scope.Registration.TransactionNo,
                    },
                    RegistrationFeeList: $scope.RegistrationFeeList
                }
                $scope.dataLoading = true;
                adminService.saveStudentRegistration(saveStudentRegistrationModel)
                    .then(function (response) {
                        if (response.data.Message == 'Success') {
                            toastr.success("Student Registration Successfully Done.");
                            $scope.resetForm();
                            window.open("/admin/PrintRegistrationReceipt/" + response.data.StudentRegistrationId);
                        } else {
                            toastr.error(response.data.Message);
                        }
                        $scope.dataLoading = false;
                    }, function (err) {
                        toastr.error("Some error occured while fetching data");
                        $scope.dataLoading = false;
                    })

            } else {
                toastr.error("Fill all fields!", "Validation Error");
            }
        }

        $scope.resetForm = function () {
            $scope.resetRegistration();
            $scope.changePermanentState();
            $scope.formRegistration.$setPristine();
            $scope.formRegistration.$setUntouched();
        }
    }
}());

File: SchoolERP/Controllers/api/StudentRegistrationController.cs
using Microsoft.Ajax.Utilities;
using School;
using SchoolERP.Models;
using System;
using System.Collections.Generic;
using System.Data.Common;
using System.Dynamic;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Web.Http;

namespace SchoolERP.Controllers.api
{
    [CheckSessionTimeOutAttribute]
    [RoutePrefix("api/StudentRegistration")]
    public class StudentRegistrationController : ApiController
    {
        [HttpPost]
        [Route("saveStudentRegistration")]
        public ExpandoObject SaveStudentRegistration(SaveStudentRegistrationModel saveStudentRegistration)
        {
            dynamic response = new ExpandoObject();
            SchoolDataContext dataContext = new SchoolDataContext();
            DbTransaction transaction = null;
            int y = 0;
            try
            {
                dataContext.Connection.Open();
                transaction = dataContext.Connection.BeginTransaction();
                dataContext.Transaction = transaction;

                //Student Registration
                var studentRegistrationModel = saveStudentRegistration.StudentRegistration;
                StudentRegistration studentRegistration = new StudentRegistration
                {
                    AadharNo = studentRegistrationModel.AadharNo,
                    AcademicSessionId = studentRegistrationModel.AcademicSessionId,
                    Address = studentRegistrationModel.Address,
                    PermanentCityId = studentRegistrationModel.PermanentCityId,
                    ClassId = studentRegistrationModel.ClassId,
                    DOB = studentRegistrationModel.DOB,
                    EmailId = studentRegistrationModel.EmailId,
                    FatherName = studentRegistrationModel.FatherName,
                    Gender = studentRegistrationModel.Gender,
                    MobileNo = studentRegistrationModel.MobileNo,
                    MotherName = studentRegistrationModel.MotherName,
                    Name = studentRegistrationModel.Name,
                    RegistrationDate = studentRegistrationModel.RegistrationDate,
                    RegistrationNo = LoadData.GenerateRegNo(dataContext, studentRegistrationModel.AcademicSessionId),
                    RegistrationStatus = (byte)RegistrationStatus.Active,
                    Remarks = studentRegistrationModel.Remarks,
                    StudentTypeId = studentRegistrationModel.StudentTypeId
                };

                dataContext.StudentRegistrations.InsertOnSubmit(studentRegistration);
                dataContext.SubmitChanges();

                //Fee Collection
                var feeCollectionModel = saveStudentRegistration.FeeCollection;
                PaymentMode paymentMode = (PaymentMode)Convert.ToByte(feeCollectionModel.PaymentMode);
                decimal LineTotal = saveStudentRegistration.RegistrationFeeList.Sum(x => x.Amount);
                FeeCollection feeCollection = new FeeCollection
                {
                    ReceiptNo = FeeData.GenerateReceiptNo(dataContext),
                    RegistrationId = studentRegistration.StudentRegistrationId,
                    AcademicSessionId = studentRegistration.AcademicSessionId,
                    PaymentDate = feeCollectionModel.PaymentDate,
                    PaymentMode = feeCollectionModel.PaymentMode,
                    SchoolBankId = null,
                    PaymentFor = (byte)PaymentFor.Registration,
                    Remarks = feeCollectionModel.Remarks,
                    RemarksForConcession = null,
                    CreatedBy = SessionUtilies.GetLoginId(),
                    CreatedOn = DateTime.Now,
                    LastUpdatedOn = DateTime.Now,
                    LastUpdatedBy = SessionUtilies.GetLoginId(),
                    FeeCollectionStatus = (byte)FeeCollectionStatus.Paid,
                    PreviousDues = 0,
                    DuesAmount = 0,
                    LineTotal = LineTotal,
                    Concession = 0,
                    Subtotal = LineTotal,
                    PaidAmount = LineTotal,
                };
                if (paymentMode == PaymentMode.Cheque)
                {
                    feeCollection.ChequeNo = feeCollectionModel.ChequeNo;
                    feeCollection.ChequeDate = feeCollectionModel.ChequeDate;
                    feeCollection.ChequeBank = feeCollectionModel.ChequeBank;
                    feeCollection.ChequeBankBranch = feeCollectionModel.ChequeBankBranch;
                }
                //else if (paymentMode == PaymentMode.DemandDraft)
                //{
                //    feeCollection.DemandDraftNo = feeCollectionModel.DemandDraftNo;
                //    feeCollection.DemandDraftDate = feeCollectionModel.DemandDraftDate;
                //    feeCollection.DemandDraftBank = feeCollectionModel.DemandDraftBank;
                //}
                else if (paymentMode == PaymentMode.UPI || paymentMode == PaymentMode.Online)
                {
                    feeCollection.TransactionNo = feeCollectionModel.TransactionNo;
                }
                dataContext.FeeCollections.InsertOnSubmit(feeCollection);
                dataContext.SubmitChanges();

                //Fee Collection detail
                saveStudentRegistration.RegistrationFeeList.ForEach(registrationFee =>
                {
                    var feeHead = (from fh in dataContext.FeeHeads
                                   where fh.FeeHeadId == registrationFee.FeeHeadId
                                   select fh).First();

                    FeeCollectionDetail feeCollectionDetail = new FeeCollectionDetail
                    {
                        FeeCollectionId = feeCollection.FeeCollectionId,
                        FeeHeadId = feeHead.FeeHeadId,
                        IsIncluded = false,
                        FeeHeadTitle = feeHead.FeeHeadTitle,
                        FeeType = (byte)FeeType.Registration,
                        Amount = registrationFee.Amount,
                        Concession = 0,
                        Total = registrationFee.Amount,
                        Comment = registrationFee.Remarks,
                        LastUpdatedOn = DateTime.Now,
                        LastUpdatedBy = SessionUtilies.GetLoginId()
                    };
                    dataContext.FeeCollectionDetails.InsertOnSubmit(feeCollectionDetail);
                    dataContext.SubmitChanges();

                });

                transaction.Commit();
                response.StudentRegistrationId = studentRegistration.StudentRegistrationId;
                response.Message = ConstantData.SuccessMessage;
            }
            catch (Exception ex)
            {
                if (y != 1)
                    transaction.Rollback();
                response.Message = ex.Message;
            }
            finally
            {
                if (null != dataContext.Connection)
                    dataContext.Connection.Close();
            }
            return response;
        }

        [HttpGet]
        [Route("StudentRegistrationList")]
        public ExpandoObject StudentRegistrationList(int AcademicSessionId)
        {
            dynamic response = new ExpandoObject();
            try
            {
                SchoolDataContext dataContext = new SchoolDataContext();
                var list = (from s1 in dataContext.StudentRegistrations
                            join as1 in dataContext.AcademicSessions
                            on s1.AcademicSessionId equals as1.AcademicSessionId
                            join c1 in dataContext.Classes on s1.ClassId equals c1.ClassId
                            join sc1 in dataContext.StudentTypes
                            on s1.StudentTypeId equals sc1.StudentTypeId
                            where s1.AcademicSessionId == AcademicSessionId
                            orderby s1.RegistrationNo descending
                            select new
                            {
                                s1.StudentRegistrationId,
                                s1.RegistrationNo,
                                s1.Name,
                                s1.DOB,
                                s1.Gender,
                                s1.FatherName,
                                s1.MotherName,
                                s1.Address,
                                s1.EmailId,
                                s1.MobileNo,
                                s1.RegistrationDate,
                                s1.Remarks,
                                c1.ClassName,
                                sc1.TypeName,
                                as1.SessionName,
                                s1.RegistrationStatus,
                                s1.City.CityName,
                                s1.PermanentCityId,
                                s1.City.StateId,
                                s1.ClassId,
                                s1.StudentTypeId,
                                s1.AcademicSessionId,
                                s1.City.State.StateName,
                                s1.AadharNo,
                            });
                response.StudentRegistrationList = list.ToList();
                response.Message = "Success";
            }
            catch (Exception ex)
            {
                response.Message = ex.Message;
            }
            return response;
        }

        [HttpGet]
        [Route("StudentRegistrationListForSearch")]
        public ExpandoObject StudentRegistrationListForSearch(byte RegistrationStatus)
        {
            dynamic response = new ExpandoObject();
            try
            {
                SchoolDataContext dataContext = new SchoolDataContext();
                var list = (from s1 in dataContext.StudentRegistrations
                            join as1 in dataContext.AcademicSessions
                            on s1.AcademicSessionId equals as1.AcademicSessionId
                            join c1 in dataContext.Classes on s1.ClassId equals c1.ClassId
                            join sc1 in dataContext.StudentTypes
                            on s1.StudentTypeId equals sc1.StudentTypeId
                            where (RegistrationStatus > 0 ? s1.RegistrationStatus == RegistrationStatus : true)
                            orderby s1.RegistrationNo descending
                            select new
                            {
                                s1.StudentRegistrationId,
                                s1.RegistrationNo,
                                SearchField = s1.RegistrationNo + " (" + s1.Name + ")",
                                s1.Name,
                                s1.DOB,
                                s1.Gender,
                                s1.FatherName,
                                s1.MotherName,
                                s1.Address,
                                s1.EmailId,
                                s1.MobileNo,
                                s1.RegistrationDate,
                                s1.Remarks,
                                c1.ClassName,
                                sc1.TypeName,
                                as1.SessionName,
                                s1.RegistrationStatus,
                                s1.City.CityName,
                                s1.PermanentCityId,
                                s1.City.StateId,
                                s1.ClassId,
                                s1.StudentTypeId,
                                s1.AcademicSessionId,
                                s1.City.State.StateName,
                                s1.AadharNo,
                            });
                response.StudentRegistrationList = list.ToList();
                response.Message = "Success";
            }
            catch (Exception ex)
            {
                response.Message = ex.Message;
            }
            return response;
        }

        [Route("UpdateStudentRegistration")]
        public ExpandoObject UpdateStudentRegistration(StudentRegistrationModel studentRegistrationModel)
        {
            dynamic response = new ExpandoObject();
            try
            {
                SchoolDataContext dataContext = new SchoolDataContext();
                StudentRegistration studentRegistration = dataContext.StudentRegistrations.Where(x => x.StudentRegistrationId == studentRegistrationModel.StudentRegistrationId).First();
                studentRegistration.AadharNo = studentRegistrationModel.AadharNo;
                studentRegistration.AcademicSessionId = studentRegistrationModel.AcademicSessionId;
                studentRegistration.Address = studentRegistrationModel.Address;
                studentRegistration.PermanentCityId = studentRegistrationModel.PermanentCityId;
                studentRegistration.ClassId = studentRegistrationModel.ClassId;
                studentRegistration.DOB = studentRegistrationModel.DOB;
                studentRegistration.EmailId = studentRegistrationModel.EmailId;
                studentRegistration.FatherName = studentRegistrationModel.FatherName;
                studentRegistration.Gender = studentRegistrationModel.Gender;
                studentRegistration.MobileNo = studentRegistrationModel.MobileNo;
                studentRegistration.MotherName = studentRegistrationModel.MotherName;
                studentRegistration.Name = studentRegistrationModel.Name;
                studentRegistration.RegistrationDate = studentRegistrationModel.RegistrationDate;
                studentRegistration.Remarks = studentRegistrationModel.Remarks;
                studentRegistration.StudentTypeId = studentRegistrationModel.StudentTypeId;
                dataContext.SubmitChanges();
                response.Message = ConstantData.SuccessMessage;
            }
            catch (Exception ex)
            {

                response.Message = ex.Message;
            }

            return response;
        }

        //[HttpGet]
        //[Route("UpdateRegistrationNo")]
        //public ExpandoObject UpdateRegistrationNo()
        //{
        //    dynamic response = new ExpandoObject();
        //    try
        //    {
        //        SchoolDataContext dataContext = new SchoolDataContext();
        //        var academicSessionIds = dataContext.AcademicSessions.Select(x => x.AcademicSessionId).ToList();

        //        foreach (var aId in academicSessionIds)
        //        {
        //            int serialNo = 1;
        //            var classNo = (from ass in dataContext.AcademicSessions
        //                           join cr in dataContext.Classes
        //                           on ass.ClassId equals cr.ClassId
        //                           where ass.AcademicSessionId == aId
        //                           select cr.ClassNo).Distinct();
        //            var registrations = dataContext.StudentRegistrations.Where(x => x.AcademicSessionId == aId).OrderBy(x => x.StudentRegistrationId).ToList();
        //            foreach (var reg in registrations)
        //            {
        //                StudentRegistration studentRegistration = (from st in dataContext.StudentRegistrations
        //                                                           where st.StudentRegistrationId == reg.StudentRegistrationId
        //                                                           select st).First();
        //                string sessionName = dataContext.GetSessionForRegNo(aId);
        //                studentRegistration.RegistrationNo = "R" + classNo.First() + dataContext.GetSessionForRegNo(aId).ToString() + serialNo.ToString("D" + 4);
        //                dataContext.SubmitChanges();
        //                serialNo++;
        //            }
        //        }

        //        response.Message = ConstantData.SuccessMessage;
        //    }
        //    catch (Exception ex)
        //    {

        //        response.Message = ex.Message;
        //    }

        //    return response;
        //}
    }
}
